src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,179 @@
+  }
+
+  def emitSampleIdentifierBlock(sampleIds: Array[String]): Array[Byte] = {
+    val sampleIdByteBuffer = new ArrayBuilder[Byte]
+    val byteBuffer = new ArrayBuilder[Byte]
+
+    sampleIds.foreach(s => stringToBytesWShortLength(sampleIdByteBuffer, s))
+
+    val sampleIdBytes = sampleIdByteBuffer.result()
+    val sampleBlockLength = sampleIdBytes.length + 8
+    val nSamples = sampleIds.length
+
+    intToBytesLE(byteBuffer, sampleBlockLength)
+    intToBytesLE(byteBuffer, nSamples)
+    sampleIdBytes.foreach(byteBuffer += _)
+
+    byteBuffer.result()
+  }
+
+  def emitGenotypeData(gs: Iterable[Genotype], nSamples: Int,
+    nAlleles: Int, nBitsPerProb: Int, conversionFactor: Double, bitMask: Long): (Array[Byte], Int, Int) = {
+    val byteBuffer = new ArrayBuilder[Byte]
+    val sampleProbs = new ArrayBuilder[Int]
+    val nGenotypes = triangle(nAlleles)
+    val noCall = new Array[Int](nGenotypes - 1)
+
+    intToBytesLE(byteBuffer, nSamples)
+    shortToBytesLE(byteBuffer, nAlleles)
+    byteBuffer += ploidy
+    byteBuffer += ploidy
+
+    gs.foreach { gt =>
+      val px = gt.unboxedPX
+      if (px == null) {
+        byteBuffer += 0x80.toByte
+        noCall.foreach(sampleProbs += _)
+      }
+      else {
+        byteBuffer += ploidy
+        resizeProbInts(px, conversionFactor).dropRight(1).foreach(sampleProbs += _)
+      }
+    }
+
+    byteBuffer += phased
+    byteBuffer += nBitsPerProb.toByte
+    bitPack(byteBuffer, sampleProbs.result(), nBitsPerProb)
+
+    val uncompressedData = byteBuffer.result()
+    val compressedData = compress(uncompressedData)
+
+    (compressedData, compressedData.length, uncompressedData.length)
+  }
+
+  def emitVariantBlock(v: Variant, va: Annotation, gs: Iterable[Genotype],
+    rsidQuery: Option[Querier], varidQuery: Option[Querier], nSamples: Int, nBitsPerProb: Int,
+    conversionFactor: Double, bitMask: Long): Array[Byte] = {
+
+    val byteBuffer = new ArrayBuilder[Byte]
+    val nAlleles = v.nAlleles
+
+    require(nAlleles <= 0xffff, s"Maximum number of alleles per variant is ${ 0xffff }. Found ${ v.nAlleles }.")
+
+    val varid = varidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(v.toString)
+    val rsid = rsidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(".")
+
+    stringToBytesWShortLength(byteBuffer, varid)
+    stringToBytesWShortLength(byteBuffer, rsid)
+    stringToBytesWShortLength(byteBuffer, v.contig)
+    intToBytesLE(byteBuffer, v.start)
+    shortToBytesLE(byteBuffer, nAlleles)
+    stringToBytesWIntLength(byteBuffer, v.ref)
+    v.altAlleles.foreach(a => stringToBytesWIntLength(byteBuffer, a.alt))
+
+    val (gtData, compressedLength, decompressedLength) = emitGenotypeData(gs, nSamples, nAlleles, nBitsPerProb, conversionFactor, bitMask)
+    intToBytesLE(byteBuffer, compressedLength + 4)
+    intToBytesLE(byteBuffer, decompressedLength)
+    gtData.foreach(byteBuffer += _)
```

Comment:
Have emitGenotypeData take a array builder instead of returning an array.  Basically, for every block of data you're going to emit, you should allocate ONE ArrayBuilder and put everything into it, and only call result() at the very end.